04_data_structures/07_roaring_bitmaps/rle.go

```diff
@@ -1,11 +1,232 @@
 package bitmap
 
+import (
+	"encoding/binary"
+	"fmt"
+)
+
+// ==========
+// First idea
+// ==========
+// Each byte contains either compressed or uncompressed data:
+// Bytes starting with 0 are uncompressed bytes, with
+// leading 0 followed by sequence of literal 0s and 1s,
+// 01001011 1000100
+//     ^^^^ ^^^^^^^ - up to 12 bits of uncompressed data
+//  ^^^             - length of uncompressed data, [1; 15]
+// ^                - uncompressed byte flag
+
+// In our scheme we start to make profit only if run is longer than 8,
+// thus we can use bits encoding the length to encode values starting from 8
+// Bytes starting with 1 are compressed bytes:
+// 10100001
+//        ^ - value of the run
+//  ^^^^^^  - length of the run, starting from 2^3 to to 2^9 - 1 = 511
+// ^        - compressed byte flag
+
+// Disadvantages:
+// annoying to work with individual bits
+//
+
+// ===========
+// Second idea
+// ===========
+// Let's work with whole bytes instead:
```

Comment:
Good idea, though it's not clear to me whether the convenience you gain is worth it. In particular, an interesting exercise could be to think about if there's any pathological cases that could really increase the memory usage.